sync15-adapter/src/state_machine.rs

```diff
@@ -0,0 +1,599 @@
+// `meta/global` and `crypto/keys`, and add some states for "needs fresh
+// token" and "has fresh token, check if endpoint changed". If thee endpoint is
+// the same, continue to `InitialWithLiveToken`; if different, go to
+// `FreshStartRequired`.
+
+// TODO: Should we expose this over FFI? I think not, we want the library to
+// automatically handle getting a new token and setting everything up so that
+// the app can just sync.
+struct SyncStateMachine<'s, 'k> {
+    client: Rc<Client>,
+    scratchpad: &'s mut Scratchpad,
+    keys: &'k KeyBundle,
+    label_sequence: Vec<&'static str>,
+}
+
+impl<'s, 'k> SyncStateMachine<'s, 'k> {
+    fn new(
+        client: Rc<Client>,
+        scratchpad: &'s mut Scratchpad,
+        keys: &'k KeyBundle,
+    ) -> SyncStateMachine<'s, 'k> {
+        SyncStateMachine {
+            client,
+            scratchpad,
+            keys,
+            label_sequence: Vec::new(),
+        }
+    }
+
+    // Runs through the state machine to the ready state.
+    fn to_ready(&mut self) -> error::Result<()> {
+        let mut state = InitialWithLiveToken;
+        loop {
+            match &state {
+                Ready => {
+                    self.label_sequence.push(&state.label());
+                    break;
+                }
+                // If we already started over once before, we're likely in a
+                // cycle, and should try again later. Like the iOS state
+                // machine, other cycles aren't a problem; we'll cycle through
+                // earlier states if we need to reupload `meta/global` or
+                // `crypto/keys`.
+                FreshStartRequired if self.label_sequence.contains(&state.label()) => {
+                    bail!(error::unexpected("State machine cycle error"))
+                }
+                _ => {
+                    self.label_sequence.push(&state.label());
+                    state = state.advance(self.client.clone(), self.scratchpad, &self.keys)?;
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+/// Whether we should skip fetching `meta/global` or `crypto/keys` from the
+/// server because our locally cached copy is up-to-date, fetch a fresh copy
+/// from the server, or invalidate cached state, then fetch.
+enum FetchAction {
+    Skip,
+    Fetch,
+    InvalidateThenFetch,
+}
+
+enum SyncState {
+    InitialWithLiveToken,
+    InitialWithLiveTokenAndInfo {
+        collections: HashMap<String, ServerTimestamp>,
+    },
+    NeedsFreshMetaGlobal {
+        collections: HashMap<String, ServerTimestamp>,
+    },
+    HasMetaGlobal {
+        collections: HashMap<String, ServerTimestamp>,
+    },
+    ResolveMetaGlobal {
+        collections: HashMap<String, ServerTimestamp>,
+        global: BsoRecord<MetaGlobalRecord>,
+    },
+    NeedsFreshCryptoKeys,
+    Ready,
+    FreshStartRequired,
+}
+
+impl SyncState {
+    fn label(&self) -> &'static str {
```

Comment:
It's pretty gross and and error prone to use strings here. Maybe we can use [std::mem::discriminant](https://doc.rust-lang.org/std/mem/fn.discriminant.html) instead?

Reply:
Oh, cool, and then I can accumulate `Discriminant`s instead of `&str`s in `label_sequence`. 📓 Thanks!